### DISCUSSION TITLE
Why can't I use pattern on query parameters whose type is List?

### DISCUSSION THREAD

**Discussion by mike-egg123** [fastapi/fastapi#12542] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Query
from typing import Union, List

app = FastAPI()

# 使用Query来对查询参数做限制
@app.get('/query/')
async def query_limit(q: Union[List[str], None] = Query(default=..., max_length=20, min_length=2, pattern='1\d{10}')):
    # default=...显式声明该参数是必须参数，也可以直接不设置默认值来声明该参数是必须参数
    return {'q': q}
```

### Description

When I try to send the request like: 'http://127.0.0.1:8000/query/?q=13900000000&q=13700000000&q=13800000000', an Internel Server Error happened. The error message is: TypeError: Unable to apply constraint 'pattern' to supplied value ['13900000000', '13700000000', '13800000000'] for schema of type 'list'. So I want to know how can I use pattern to check each 'q'?

### Operating System

Windows

### FastAPI Version

0.115.3

### Pydantic Version

2.9.2

### Python Version

3.9.20

**Comment by 1001pepi** [comment]
You can use the pydantic StringContraints like this:

```python
from fastapi import FastAPI, Query
from typing import Annotated, List
from pydantic import BaseModel, StringConstraints


class QParam(BaseModel):
    values: List[
        Annotated[
            str,
            StringConstraints(
                max_length=20, min_length=2, pattern='1\d{10}'
            ),
        ]
    ] | None = None

app = FastAPI()


# 使用Query来对查询参数做限制
@app.get('/query/')
async def query_limit(q: Annotated[QParam, Query()]):
    # default=...显式声明该参数是必须参数，也可以直接不设置默认值来声明该参数是必须参数
    return {'q': q.values}
```

**Comment by YuriiMotov** [comment, marked as answer]
You need to specify pattern for list item, but in your code you specify this parameter for `list`.

You can create the alias for type, specify its constraints, and use this type as the type of list items: 

`MyStr: TypeAlias = Annotated[str, Field(pattern=r"1\d{10}")]`


```
from fastapi import FastAPI, Query
from typing import Annotated, TypeAlias, Union, List

from pydantic import Field

app = FastAPI()

MyStr: TypeAlias = Annotated[str, Field(pattern=r"1\d{10}")]

@app.get("/query/")
async def query_limit(
    q: Union[List[MyStr], None] = Query(default=..., max_length=20, min_length=2),

    # Or, version with Annotated:
    # q: Annotated[list[MyStr] | None, Query(max_length=20, min_length=2)],
):
    return {"q": q}
```

**PS**: I assumed that the intention of using `max_length` and `min_length` was to limit array size, because the length of list elements is already limited by regexp and it would be strange to limit it twice with different limits.
